File: SimpleWebApp.WebApi/Dtos/InventoryDto.cs
using System.ComponentModel.DataAnnotations;
using SimpleWebApp.WebApi.Enumerations;

namespace SimpleWebApp.WebApi.Dtos;

public sealed class InventoryDto
{
    private string? _itemName;
    private string? _warehouseLocation;

    public long Id { get; set; }
    public string ItemName
    {
        get => _itemName!;
        set
        {
            ArgumentNullException.ThrowIfNull(value);
            _itemName = value;
        }
    }

    public string WarehouseLocation
    {
        get => _warehouseLocation!;
        set
        {
            ArgumentNullException.ThrowIfNull(value);
            _warehouseLocation = value;
        }
    }

    public string? Description { get; set; }
    public GoodType GoodType { get; set; }

    [Required]
    public long ItemId { get; set; }

    [Required]
    public long WarehouseId { get; set; }
    
    [Required]
    [Range(1, long.MaxValue, ErrorMessage = "Quantity should be positive integer greater than 0")]
    public long Quantity { get; set; }
}

File: VueClient/src/models/inventory-dto.ts
import { GoodType } from "./good-type";

export default class InventoryDto {
  public id: number;
  public itemName: string;
  public description: string;
  public goodType: GoodType;
  public goodTypeString?: string;
  public itemId: number;
  public warehouseId: number;
  public warehouseLocation: string;
  public quantity: number;

  constructor(
    id?: number,
    itemName?: string,
    description?: string,
    goodType?: GoodType,
    itemId?: number,
    warehouseId?: number,
    warehouseLocation?: string,
    quantity?: number
  ) {
    this.id = id || 0;
    this.itemName = itemName || "";
    this.description = description || "";
    this.goodType = goodType || GoodType.General;
    this.itemId = itemId || -1;
    this.warehouseId = warehouseId || -1;
    this.warehouseLocation = warehouseLocation || "";
    this.quantity = quantity || 0;
  }

  deserialize(inventory: InventoryDto): InventoryDto {
    if (inventory) {
      Object.assign(this, inventory);
      this.goodTypeString = GoodType[this.goodType];
    }
    return this;
  }
}
